### ISSUE TITLE
Support HTTP  in Driver

### ISSUE THREAD

**Issue by gingerwizard** [ClickHouse/clickhouse-go#597]

**Comment by ortyomka** [comment]
Hello, @gingerwizard. I want to clarify this issue.

Does it mean to add HTTP-interface? So, dev could choose Native, Std or HTTP to use.

or

Expand Native Driver to accept both Native and HTTP interfaces? So, dev write addresses with prefix protocol(http://foo:port and clickhouse://foo:port) and driver will select interface by its own.

**Comment by futuarmo** [comment]
@gingerwizard, hello! Is it appropriate to adapt https://github.com/mailru/go-clickhouse implementation?

**Comment by gingerwizard** [comment]
@ortyomka has started this work for the std interface. @ortyomka, ideally we would continue to send the data in native format since this will give us the best performance. Happy to share the work or let you own.

**Comment by ortyomka** [comment]
@gingerwizard Do I get it right that you want to pass requests in native format via HTTP Interface?

**Comment by crashiura** [comment]
@gingerwizard I have another solution, you can see it please, if the solution is good, I will finalize it

**Comment by gingerwizard** [comment]
@ortyomka 

> Do I get it right that you want to pass requests in native format via HTTP Interface?

yes

**Comment by ortyomka** [comment]
@gingerwizard 
Interesting, I didn't see this option in the documentation, but I'll research it.

If you already know how to do this, I would be happy to see an example.

**Comment by crashiura** [comment]
@gingerwizard ok i will raise PR

**Comment by gingerwizard** [comment]
Discussed with @genzgd 

1. We prefer the usage pattern of @ortyomka approach for the user 
2. httpBatch is a nice improvement from @crashiura 
3. `type transport interface {` is a bit ugly..maybe @crashiura idea of a batch would allow this?

Lets proceed with @ortyomka with the batch approach and see if it cleanly separates.

**Comment by ortyomka** [comment]
@crashiura Yes, I will

**Comment by gingerwizard** [comment]
@ortyomka can you summarize current state here and remanining TODOs.

**Comment by ortyomka** [comment]
Now:
- HTTP supported by std driver
- HTTP works with native format queries
- HTTP support Batch, Exec, Query

TODO:
- HTTP URL options (like "query_id")
- HTTP Async Insert
- HTTP support by clickhouse driver V2

This is how I see this, and I will open PR soon

cc @gingerwizard

**Comment by gingerwizard** [comment]
Data race in new http code @ortyomka 

```
make test
==================
WARNING: DATA RACE
Write at 0x00c00011bad0 by goroutine 77:
  github.com/ClickHouse/clickhouse-go/v2.(*httpBatch).Send.func2()
      /opt/clickhouse-go/conn_http_batch.go:164 +0x200

Previous write at 0x00c00011bad0 by goroutine 53:
  github.com/ClickHouse/clickhouse-go/v2.(*httpBatch).Send()
      /opt/clickhouse-go/conn_http_batch.go:179 +0x49c
  github.com/ClickHouse/clickhouse-go/v2/lib/driver.Batch.Send-fm()
      <autogenerated>:1 +0x48
  github.com/ClickHouse/clickhouse-go/v2.(*stdDriver).Commit()
      /opt/clickhouse-go/clickhouse_std.go:152 +0x8c
  database/sql.(*Tx).Commit.func1()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/database/sql/sql.go:2290 +0x54
  database/sql.withLock()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/database/sql/sql.go:3437 +0x88
  database/sql.(*Tx).Commit()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/database/sql/sql.go:2289 +0x184
  github.com/ClickHouse/clickhouse-go/v2/tests/std.TestStdMap.func1()
      /opt/clickhouse-go/tests/std/map_test.go:76 +0x560
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/testing/testing.go:1439 +0x18c
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/testing/testing.go:1486 +0x44

Goroutine 77 (running) created at:
  github.com/ClickHouse/clickhouse-go/v2.(*httpBatch).Send()
      /opt/clickhouse-go/conn_http_batch.go:159 +0x37c
  github.com/ClickHouse/clickhouse-go/v2/lib/driver.Batch.Send-fm()
      <autogenerated>:1 +0x48
  github.com/ClickHouse/clickhouse-go/v2.(*stdDriver).Commit()
      /opt/clickhouse-go/clickhouse_std.go:152 +0x8c
  database/sql.(*Tx).Commit.func1()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/database/sql/sql.go:2290 +0x54
  database/sql.withLock()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/database/sql/sql.go:3437 +0x88
  database/sql.(*Tx).Commit()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/database/sql/sql.go:2289 +0x184
  github.com/ClickHouse/clickhouse-go/v2/tests/std.TestStdMap.func1()
      /opt/clickhouse-go/tests/std/map_test.go:76 +0x560
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/testing/testing.go:1439 +0x18c
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/testing/testing.go:1486 +0x44

Goroutine 53 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/testing/testing.go:1486 +0x560
  github.com/ClickHouse/clickhouse-go/v2/tests/std.TestStdMap()
      /opt/clickhouse-go/tests/std/map_test.go:32 +0x168
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/testing/testing.go:1439 +0x18c
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.18.3/libexec/src/testing/testing.go:1486 +0x44
==================
```

**Comment by ortyomka** [comment]
I will fix

**Comment by ortyomka** [comment]
Fixed, https://github.com/ClickHouse/clickhouse-go/pull/642#issuecomment-1172676944

**Comment by ortyomka** [comment]
This PR #642 is the last one, it's small, most of the lines are tests.

@gingerwizard , would you be able to check it out?

I would like to finish this task this week.

**Comment by ortyomka** [comment]
@gingerwizard Closed?

**Comment by gingerwizard** [comment]
I think this is done. They are some enhancement e.g. compression and HTTP streaming (right now we read the body all at once) but I think we can close.
